alpenhorn/acquisition.py

```diff
@@ -1,157 +1,263 @@
+    info_config : string or None
+        An optional JSON blob containing configuration data for the
+        associated FileInfo class.
+
+    Notes
+    -----
+    If `info_class` is None, then a subclass of `alpenhorn.info_base._NoInfo`
+    is used as a stand-in for the type's Info class.  This stand-in can
+    be used for type detection by specifying one or more `patterns` in the
+    `info_config` field.  This allows types to be supported by alpenhorn
+    without the need for any outside Info class implementation.
     """
 
-    name = pw.CharField(max_length=64)
+    name = pw.CharField(max_length=64, unique=True)
+    priority = pw.IntegerField(default=0, null=False)
+    info_class = pw.CharField(max_length=254, null=True)
     notes = pw.TextField(null=True)
+    info_config = pw.TextField(null=True)
 
-    # This dict is an atribute on the class, used to hold the set of registered
-    # handlers. Store as a dictionary for easy lookup of handlers by name.
-    _registered_acq_types = {}
-
-    @classmethod
-    def register_type(cls, acq_info):
-        """Register a new acquisition type with alpenhorn.
-
-        This creates the entry in the AcqType table if it does not already exist.
+    # This dict is an attribute on the class, used to hold the loaded info
+    # classes.  Store as a dictionary for easy lookup of info classes by name.
+    _info_classes = dict()
 
-        Parameters
-        ----------
-        acq_info : AcqInfoBase
-            AcqInfoBase describing the type of acquisition.
-        """
-
-        try:
-            cls.get(name=acq_info._acq_type)
-        except pw.DoesNotExist:
-            log.info('Create AcqType entry for "%s"' % acq_info._acq_type)
-            cls.create(name=acq_info._acq_type)
-
-        # Add to registry
-        cls._registered_acq_types[acq_info._acq_type] = acq_info
-
-    @classmethod
-    def check_registration(cls):
-        """Check that all AcqTypes known to the database have a
-        registered handler.
-        """
-
-        # Get the list of types names from the database
-        db_acq_types = [row[0] for row in cls.select(cls.name).tuples()]
+    class Meta:
+        indexes = ((("priority",), False),)  # index speeds ordering
 
-        # Get the names of all the registered types
-        reg_acq_types = cls._registered_acq_types.keys()
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
 
-        # Find any missing types
-        missing_acq_types = set(db_acq_types) - set(reg_acq_types)
+        # Get the import_error action for this type
+        self._info_import_errors = config.info_import_errors(
+            self.name, is_acq=isinstance(self, AcqType)
+        )
 
-        if len(missing_acq_types):
-            raise RuntimeError(
-                "AcqTypes %s have no registered handler." % repr(missing_acq_types)
-            )
+    def _get_info_class(self, acqtype: AcqType | None, is_acq: bool) -> info_base:
```

Comment:
This function and the corresponding `_get_io_class` from the previous PR do essentially the same thing:
* Fetch the value of `info_class` (or `io_class`) from the database record
* Import the module that it specifies
* Return the named class from the imported module

This works fine as-is, and gives alpenhorn broad flexibility to deal with new modules/classes being introduced to DB records after start-up.

However, I've though up another possibility that I think would do okay:
* restrict the `info_class` field to contain only a class name (or, in the case of `io_class`, the "base name" for the I/O class, like "Nearline" or "Default").
* Require third party module which implement an Info Class or I/O class to appear in the extensions list
* The `register_extension` function of one of these classes needs to return a list of InfoClasses and/or I/O modules which they implement.
* InfoClasses declared by `register_extension` get added to a container module `alpenhorn.info` and the class names found in the `info_class` for Acq/FileTypes then refer to classes in that namespace.
* Similarly I/O modules are added to `alpenhorn.io`.

This has the benefit that all module import happens at start-up.  And it also removes the possibility for arbitrary code execution via database injection attack.  However, something would need to be figured out with how to deal with classes which are declared in the DB but not loaded from an extension.